ops/terraform/services/pipeline/modules/bfd_pipeline_slis/lambda_src/update_pipeline_slis.py

```diff
@@ -0,0 +1,629 @@
+        event_timestamp = datetime.fromisoformat(event_time_iso.removesuffix("Z"))
+    except KeyError as exc:
+        print(f'Record did not contain any key with "eventTime": {exc}')
+        return
+    except ValueError as exc:
+        print(f"Event timestamp was not in valid ISO format: {exc}")
+        return
+
+    try:
+        file_key = record["s3"]["object"]["key"]
+    except KeyError as exc:
+        print(f"No bucket file found in event notification: {exc}")
+        return
+
+    decoded_file_key = unquote(file_key)
+    status_group_str = "|".join([e.value for e in PipelineDataStatus])
+    rif_types_group_str = "|".join([e.value for e in RifFileType])
+    # The incoming file's key should match an expected format, as follows:
+    # "<Incoming/Done>/<ISO date format>/<file name>".
+    if match := re.search(
+        rf"^({status_group_str})/([\d\-:TZ]+)/.*({rif_types_group_str}).*$",
+        decoded_file_key,
+        re.IGNORECASE,
+    ):
+        pipeline_data_status = PipelineDataStatus(match.group(1).lower())
+        ccw_timestamp = match.group(2)
+        rif_file_type = RifFileType(match.group(3).lower())
+
+        rif_type_dimension = {"data_type": rif_file_type.name.lower()}
+        group_timestamp_dimension = {"group_timestamp": ccw_timestamp}
+
+        timestamp_metric_name = f"time/data-{pipeline_data_status.name.lower()}"
+
+        utc_timestamp = calendar.timegm(event_timestamp.utctimetuple())
+
+        # An inline function is defined here to pass to backoff_retry() as Python does not support
+        # multiple line lambdas, so this is the next-best option
+        def put_timestamp_metrics():
+            # Store four metrics (gen_all_dimensioned_metrics() will generate all possible
+            # dimensions of the given metric based upon the powerset of the dimensions):
+            put_metric_data(
+                metric_namespace=METRICS_NAMESPACE,
+                metrics=gen_all_dimensioned_metrics(
+                    metric_name=timestamp_metric_name,
+                    timestamp=event_timestamp,
+                    value=utc_timestamp,
+                    unit="Seconds",
+                    dimensions=[rif_type_dimension, group_timestamp_dimension],
+                ),
+            )
+
+        try:
+            print(
+                f'Putting data timestamp metrics "{METRICS_NAMESPACE}/{timestamp_metric_name}" up'
+                f" to CloudWatch with unix timestamp value {utc_timestamp}"
+            )
+            backoff_retry(
+                func=put_timestamp_metrics, ignored_exceptions=COMMON_UNRECOVERABLE_EXCEPTIONS
+            )
+            print(f'Successfully put metrics to "{METRICS_NAMESPACE}/{timestamp_metric_name}"')
+        except Exception as exc:
+            print(
+                "An unrecoverable error occurred when trying to call PutMetricData for metric"
+                f" {METRICS_NAMESPACE}/{timestamp_metric_name}: {exc}"
+            )
```

Comment:
If this is unrecoverable, do we want to return here? (We do in some other unrecoverable scenarios below)

Reply:
Good catch, I had forgotten to include the `return` within the catch block. I've added the `return` back.